File: src/extras-kparts/libkde_terminal_interface.cpp
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVariant>
#include <kde_terminal_interface.h>
#include "libkde_terminal_interface.h"
#include "libkde_terminal_interface.hxx"

void TerminalInterface_StartProgram(TerminalInterface* self, const libqt_string program, const libqt_list /* of libqt_string */ args) {
    QString program_QString = QString::fromUtf8(program.data, program.len);
    QList<QString> args_QList;
    args_QList.reserve(args.len);
    libqt_string* args_arr = static_cast<libqt_string*>(args.data);
    for (size_t i = 0; i < args.len; ++i) {
        QString args_arr_i_QString = QString::fromUtf8(args_arr[i].data, args_arr[i].len);
        args_QList.push_back(args_arr_i_QString);
    }
    self->startProgram(program_QString, args_QList);
}

void TerminalInterface_ShowShellInDir(TerminalInterface* self, const libqt_string dir) {
    QString dir_QString = QString::fromUtf8(dir.data, dir.len);
    self->showShellInDir(dir_QString);
}

void TerminalInterface_SendInput(TerminalInterface* self, const libqt_string text) {
    QString text_QString = QString::fromUtf8(text.data, text.len);
    self->sendInput(text_QString);
}

int TerminalInterface_TerminalProcessId(TerminalInterface* self) {
    return self->terminalProcessId();
}

int TerminalInterface_ForegroundProcessId(TerminalInterface* self) {
    return self->foregroundProcessId();
}

libqt_string TerminalInterface_ForegroundProcessName(TerminalInterface* self) {
    QString _ret = self->foregroundProcessName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string TerminalInterface_CurrentWorkingDirectory(const TerminalInterface* self) {
    QString _ret = self->currentWorkingDirectory();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_list /* of libqt_string */ TerminalInterface_AvailableProfiles(const TerminalInterface* self) {
    QList<QString> _ret = self->availableProfiles();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QString _lv_ret = _ret[i];
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_b = _lv_ret.toUtf8();
        libqt_string _lv_str;
        _lv_str.len = _lv_b.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_b.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_string TerminalInterface_CurrentProfileName(const TerminalInterface* self) {
    QString _ret = self->currentProfileName();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

bool TerminalInterface_SetCurrentProfile(TerminalInterface* self, const libqt_string profileName) {
    QString profileName_QString = QString::fromUtf8(profileName.data, profileName.len);
    return self->setCurrentProfile(profileName_QString);
}

QVariant* TerminalInterface_ProfileProperty(const TerminalInterface* self, const libqt_string profileProperty) {
    QString profileProperty_QString = QString::fromUtf8(profileProperty.data, profileProperty.len);
    return new QVariant(self->profileProperty(profileProperty_QString));
}

void TerminalInterface_OperatorAssign(TerminalInterface* self, const TerminalInterface* param1) {
    self->operator=(*param1);
}

void TerminalInterface_Delete(TerminalInterface* self) {
    delete self;
}

File: include/extras-kparts/libkde_terminal_interface.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://api.kde.org/terminalinterface.html
pub const terminalinterface = struct {
    /// [Qt documentation](https://api.kde.org/terminalinterface.html#startProgram)
    ///
    /// ``` self: QtC.TerminalInterface, program: []const u8, args: [][]const u8, allocator: std.mem.Allocator ```
    pub fn StartProgram(self: ?*anyopaque, program: []const u8, args: [][]const u8, allocator: std.mem.Allocator) void {
        const program_str = qtc.libqt_string{
            .len = program.len,
            .data = program.ptr,
        };
        var args_arr = allocator.alloc(qtc.libqt_string, args.len) catch @panic("terminalinterface.StartProgram: Memory allocation failed");
        defer allocator.free(args_arr);
        for (args, 0..args.len) |item, i| {
            args_arr[i] = .{
                .len = item.len,
                .data = item.ptr,
            };
        }
        const args_list = qtc.libqt_list{
            .len = args.len,
            .data = args_arr.ptr,
        };
        qtc.TerminalInterface_StartProgram(@ptrCast(self), program_str, args_list);
    }

    /// [Qt documentation](https://api.kde.org/terminalinterface.html#showShellInDir)
    ///
    /// ``` self: QtC.TerminalInterface, dir: []const u8 ```
    pub fn ShowShellInDir(self: ?*anyopaque, dir: []const u8) void {
        const dir_str = qtc.libqt_string{
            .len = dir.len,
            .data = dir.ptr,
        };
        qtc.TerminalInterface_ShowShellInDir(@ptrCast(self), dir_str);
    }

    /// [Qt documentation](https://api.kde.org/terminalinterface.html#sendInput)
    ///
    /// ``` self: QtC.TerminalInterface, text: []const u8 ```
    pub fn SendInput(self: ?*anyopaque, text: []const u8) void {
        const text_str = qtc.libqt_string{
            .len = text.len,
            .data = text.ptr,
        };
        qtc.TerminalInterface_SendInput(@ptrCast(self), text_str);
    }

    /// [Qt documentation](https://api.kde.org/terminalinterface.html#terminalProcessId)
    ///
    /// ``` self: QtC.TerminalInterface ```
    pub fn TerminalProcessId(self: ?*anyopaque) i32 {
        return qtc.TerminalInterface_TerminalProcessId(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/terminalinterface.html#foregroundProcessId)
    ///
    /// ``` self: QtC.TerminalInterface ```
    pub fn ForegroundProcessId(self: ?*anyopaque) i32 {
        return qtc.TerminalInterface_ForegroundProcessId(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/terminalinterface.html#foregroundProcessName)
    ///
    /// ``` self: QtC.TerminalInterface, allocator: std.mem.Allocator ```
    pub fn ForegroundProcessName(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.TerminalInterface_ForegroundProcessName(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("terminalinterface.ForegroundProcessName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/terminalinterface.html#currentWorkingDirectory)
    ///
    /// ``` self: QtC.TerminalInterface, allocator: std.mem.Allocator ```
    pub fn CurrentWorkingDirectory(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.TerminalInterface_CurrentWorkingDirectory(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("terminalinterface.CurrentWorkingDirectory: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/terminalinterface.html#availableProfiles)
    ///
    /// ``` self: QtC.TerminalInterface, allocator: std.mem.Allocator ```
    pub fn AvailableProfiles(self: ?*anyopaque, allocator: std.mem.Allocator) [][]const u8 {
        const _arr: qtc.libqt_list = qtc.TerminalInterface_AvailableProfiles(@ptrCast(self));
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]const u8, _arr.len) catch @panic("terminalinterface.AvailableProfiles: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("terminalinterface.AvailableProfiles: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/terminalinterface.html#currentProfileName)
    ///
    /// ``` self: QtC.TerminalInterface, allocator: std.mem.Allocator ```
    pub fn CurrentProfileName(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.TerminalInterface_CurrentProfileName(@ptrCast(self));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("terminalinterface.CurrentProfileName: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/terminalinterface.html#setCurrentProfile)
    ///
    /// ``` self: QtC.TerminalInterface, profileName: []const u8 ```
    pub fn SetCurrentProfile(self: ?*anyopaque, profileName: []const u8) bool {
        const profileName_str = qtc.libqt_string{
            .len = profileName.len,
            .data = profileName.ptr,
        };
        return qtc.TerminalInterface_SetCurrentProfile(@ptrCast(self), profileName_str);
    }

    /// [Qt documentation](https://api.kde.org/terminalinterface.html#profileProperty)
    ///
    /// ``` self: QtC.TerminalInterface, profileProperty: []const u8 ```
    pub fn ProfileProperty(self: ?*anyopaque, profileProperty: []const u8) QtC.QVariant {
        const profileProperty_str = qtc.libqt_string{
            .len = profileProperty.len,
            .data = profileProperty.ptr,
        };
        return qtc.TerminalInterface_ProfileProperty(@ptrCast(self), profileProperty_str);
    }

    /// [Qt documentation](https://api.kde.org/terminalinterface.html#operator-eq)
    ///
    /// ``` self: QtC.TerminalInterface, param1: QtC.TerminalInterface ```
    pub fn OperatorAssign(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.TerminalInterface_OperatorAssign(@ptrCast(self), @ptrCast(param1));
    }

    /// [Qt documentation](https://api.kde.org/terminalinterface.html#dtor.TerminalInterface)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.TerminalInterface ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.TerminalInterface_Delete(@ptrCast(self));
    }
};
